presto-main/src/test/java/com/facebook/presto/sql/planner/PlanAssert.java

```diff
@@ -0,0 +1,475 @@
+            {
+                if (super.matches(node, symbolAliases)) {
+                    SemiJoinNode semiJoinNode = (SemiJoinNode) node;
+                    symbolAliases.put(sourceSymbolAlias, semiJoinNode.getSourceJoinSymbol());
+                    symbolAliases.put(filteringSymbolAlias, semiJoinNode.getFilteringSourceJoinSymbol());
+                    symbolAliases.put(outputAlias, semiJoinNode.getSemiJoinOutput());
+
+                    return true;
+                }
+                return false;
+            }
+        };
+    }
+
+    public static PlanMatcher joinNode(List<AliasPair> expectedEquiCriteria, PlanMatcher... sources)
+    {
+        return new NodeClassPlanMatcher(JoinNode.class, ImmutableList.copyOf(sources))
+        {
+            @Override
+            public boolean matches(PlanNode node, SymbolAliases symbolAliases)
+            {
+                if (super.matches(node, symbolAliases)) {
+                    JoinNode joinNode = (JoinNode) node;
+                    if (joinNode.getCriteria().size() == expectedEquiCriteria.size()) {
+                        int i = 0;
+                        for (JoinNode.EquiJoinClause equiJoinClause : joinNode.getCriteria()) {
+                            // TODO this does not work as joinNode.getCriteria() order is not deterministic
+                            AliasPair expectedEquiClause = expectedEquiCriteria.get(i++);
+                            symbolAliases.put(expectedEquiClause.left, equiJoinClause.getLeft());
+                            symbolAliases.put(expectedEquiClause.right, equiJoinClause.getRight());
+                        }
+                        return true;
+                    }
+                }
+                return false;
+            }
+        };
+    }
+
+    public static AliasPair aliasPair(String left, String right)
+    {
+        return new AliasPair(left, right);
+    }
+
+    public static PlanMatcher filterNode(String predicate, PlanMatcher... sources)
+    {
+        final Expression expectedPredicate = new SqlParser().createExpression(predicate);
+        return new NodeClassPlanMatcher(FilterNode.class, ImmutableList.copyOf(sources))
+        {
+            @Override
+            public boolean matches(PlanNode node, SymbolAliases symbolAliases)
+            {
+                if (super.matches(node, symbolAliases)) {
+                    FilterNode filterNode = (FilterNode) node;
+                    Expression actualPredicate = filterNode.getPredicate();
+
+                    return new ExpressionVerifier(symbolAliases).process(actualPredicate, expectedPredicate);
+                }
+                return false;
+            }
+        };
+    }
+
+    private static class ExpressionVerifier
+            extends AstVisitor<Boolean, Expression>
+    {
+        private final SymbolAliases symbolAliases;
+
+        public ExpressionVerifier(SymbolAliases symbolAliases)
```

Comment:
This approach requires to reimplement slightly modified `equals` for each `Expression`, which is redundant work.
Another approach would be to:
1. Gather `listA` of symbol names from actual list (easy visitor)
2. Gather `listB` of symbols from expected list
3. Assert that both lists have the same length.
4. Add binding `listB[i]->listA[i]` for each symbol if symbol `listB[i]` is not bound already
5. Rewrite `QualifiedNameReference` from actual tree according to bindings
6. Compare expressions tree (they both should be the same).

This assumes that once symbol is bound to name, it should always be referenced by this same in expected trees.